File: 08.claudecode/ruby/agent.rb
# frozen_string_literal: true

# agent.rb — the agent loop / REPL (Ruby port).
#
# The educational analog of Claude Code's main loop (../docs/02.architecture.md):
#   gather context -> ask model -> stream text
#     -> if it requested tools: gate each, run allowed ones, feed results back, recur
#     -> else: turn done
#
# Wires together the four subsystems: Backend (llm.rb), ToolRegistry (tools.rb),
# PermissionGate (permissions.rb), ContextManager (context.rb). Implements
# streaming-style incremental output and parallel read-only tool execution
# (read-only calls run on threads, mutating ones serially). Clean-room code.

require_relative "llm"
require_relative "tools"
require_relative "permissions"
require_relative "context"

module MiniAgent
  DEFAULT_SYSTEM = <<~SYS.strip
    You are a concise coding agent. Use the available tools to inspect and modify
    the workspace. Plan briefly, act with tools, verify your work, then stop.
    Prefer dedicated tools (read_file/write_file/edit_file) over bash for files.
  SYS

  # Structured record of one agent turn, useful for tests and transcripts.
  TurnLog = Struct.new(:text, :tool_calls, :results, :done, keyword_init: true)

  # Drives the gather->plan->act->observe loop to completion.
  class Agent
    attr_reader :turns, :context

    def initialize(backend: nil, tools: nil, gate: nil, context: nil,
                   system: DEFAULT_SYSTEM, out: $stdout, max_turns: 20, stream: true)
      @backend = backend || MockBackend.new
      @tools = tools || ToolRegistry.new(".")
      @gate = gate || PermissionGate.new
      @context = context || ContextManager.new(system: system)
      @out = out
      @max_turns = max_turns
      @stream = stream
      @turns = []
    end

    # Run the loop on `task`; return the final assistant text.
    def run(task)
      @context.add_user(task)
      final_text = ""

      @max_turns.times do
        step = ask_backend
        final_text = step.text
        @context.add_assistant(step.text, step.tool_calls)

        if step.tool_calls.empty?
          @turns << TurnLog.new(text: step.text, tool_calls: [], results: [], done: true)
          break
        end

        results = run_tools(step.tool_calls)
        @turns << TurnLog.new(text: step.text, tool_calls: step.tool_calls, results: results, done: step.done)

        @context.add_tool_results(
          step.tool_calls.zip(results).map { |call, (_, res)| [call.id, res.content, res.is_error] }
        )

        emit("\n[context compacted]\n") if @context.maybe_compact
      end

      final_text
    end

    private

    def emit(text)
      @out.write(text)
      @out.flush
    end

    def ask_backend
      messages = @context.to_api_messages
      schemas = @tools.schemas
      emit("\nassistant> ")
      if @stream
        @backend.stream(messages, schemas) { |chunk| emit(chunk) }
        step = @backend.last_step
      else
        step = @backend.step(messages, schemas)
        emit(step.text)
      end
      emit("\n")
      step
    end

    # Gate then run each call. Read-only calls run in parallel on threads.
    def run_tools(calls)
      results = {}
      parallel = []
      serial = []
      calls.each_with_index do |call, i|
        tool = @tools.get(call.name)
        (tool&.read_only ? parallel : serial) << [i, call]
      end

      threads = parallel.map do |i, call|
        Thread.new { results[i] = gate_and_run(call) }
      end
      threads.each(&:join)

      serial.each { |i, call| results[i] = gate_and_run(call) }

      (0...calls.length).map { |i| results[i] }
    end

    def gate_and_run(call)
      tool = @tools.get(call.name)
      read_only = tool ? tool.read_only : false
      allowed, reason = @gate.check(call.name, call.args, read_only: read_only)
      emit("  · #{call.name}(#{fmt_args(call.args)}) — #{reason}\n")
      unless allowed
        return [call.name, ToolResult.new(content: "Permission denied: #{reason}", is_error: true)]
      end

      result = @tools.call(call.name, call.args)
      preview = result.content.to_s.lines.first.to_s.strip
      emit("    #{preview[0, 100]}\n")
      [call.name, result]
    end

    def fmt_args(args)
      args.map { |k, v| "#{k}=#{v.to_s.gsub("\n", '\\n')[0, 40]}" }.join(", ")
    end
  end

  # A minimal interactive loop: read a task, run it, repeat.
  def self.repl(agent)
    puts "mini coding agent — type a task, or 'quit' to exit."
    loop do
      print "\nyou> "
      task = ($stdin.gets || "").strip
      break if task.empty? && $stdin.eof?
      break if %w[quit exit].include?(task.downcase)

      agent.run(task) unless task.empty?
    end
  end
end

if __FILE__ == $PROGRAM_NAME
  require "tmpdir"
  Dir.mktmpdir do |d|
    agent = MiniAgent::Agent.new(
      backend: MiniAgent::MockBackend.new,
      tools: MiniAgent::ToolRegistry.new(d),
      gate: MiniAgent::PermissionGate.new(MiniAgent::PermissionConfig.new, MiniAgent::AUTO_APPROVE)
    )
    MiniAgent.repl(agent)
  end
end

File: 08.claudecode/python/agent.py
"""agent.py — the agent loop / REPL.

This is the heart of the mini coding agent and the educational analog of Claude
Code's main loop (``docs/02.architecture.md``):

    gather context -> ask the model -> stream its text
        -> if it requested tools:
               gate each call (permissions)
               run the allowed ones
               feed results back
           and repeat
        -> else: the turn is done

It wires together the four subsystems:

* :class:`~llm.Backend`        — produces the next step (mock / real Claude / Ollama)
* :class:`~tools.ToolRegistry` — executes tool calls
* :class:`~permissions.PermissionGate` — allow/ask/deny before each call
* :class:`~context.ContextManager`     — history + token-budget compaction

It also implements **streaming-style incremental output** and **parallel
read-only tool execution** (read-only calls in the same turn run concurrently,
mutating ones run serially — mirroring how Claude Code schedules parallel-safe
tools). All original, clean-room code.
"""

from __future__ import annotations

import sys
from concurrent.futures import ThreadPoolExecutor
from dataclasses import dataclass
from typing import Any, Callable, TextIO

from context import ContextManager
from llm import Backend, MockBackend, Step, ToolCall
from permissions import PermissionGate
from tools import ToolRegistry, ToolResult


DEFAULT_SYSTEM = (
    "You are a concise coding agent. Use the available tools to inspect and "
    "modify the workspace. Plan briefly, act with tools, verify your work, then "
    "stop. Prefer dedicated tools (read_file/write_file/edit_file) over bash for "
    "file operations."
)


@dataclass
class TurnLog:
    """Structured record of one agent turn, useful for tests and transcripts."""

    text: str
    tool_calls: list[ToolCall]
    results: list[tuple[str, ToolResult]]   # (tool_name, result)
    done: bool


class Agent:
    """Drives the gather→plan→act→observe loop to completion."""

    def __init__(
        self,
        backend: Backend | None = None,
        tools: ToolRegistry | None = None,
        gate: PermissionGate | None = None,
        context: ContextManager | None = None,
        system: str = DEFAULT_SYSTEM,
        out: TextIO | None = None,
        max_turns: int = 20,
        stream: bool = True,
    ) -> None:
        self.backend = backend or MockBackend()
        self.tools = tools or ToolRegistry(".")
        self.gate = gate or PermissionGate()
        self.context = context or ContextManager(system=system)
        self.out = out or sys.stdout
        self.max_turns = max_turns
        self.stream = stream
        self.turns: list[TurnLog] = []

    # -- output helper ------------------------------------------------------

    def _emit(self, text: str, end: str = "") -> None:
        self.out.write(text + end)
        self.out.flush()

    # -- public entry point -------------------------------------------------

    def run(self, task: str) -> str:
        """Run the loop on ``task`` and return the final assistant text."""
        self.context.add_user(task)
        final_text = ""

        for _turn in range(self.max_turns):
            step = self._ask_backend()
            final_text = step.text

            # Record the assistant turn (text + any tool_use blocks).
            self.context.add_assistant(step.text, step.tool_calls)

            if not step.tool_calls:
                # No tools requested → the turn (and run) is finished.
                self.turns.append(TurnLog(step.text, [], [], True))
                break

            # Execute the requested tools, gated by permissions.
            results = self._run_tools(step.tool_calls)
            self.turns.append(TurnLog(step.text, step.tool_calls, results, step.done))

            # Feed results back as tool_result blocks for the next turn.
            self.context.add_tool_results([
                (call.id, res.content, res.is_error)
                for call, (_, res) in zip(step.tool_calls, results)
            ])

            # Manage the token budget between turns.
            if self.context.maybe_compact():
                self._emit("\n[context compacted]\n")

        return final_text

    # -- one model turn -----------------------------------------------------

    def _ask_backend(self) -> Step:
        messages = self.context.to_api_messages()
        schemas = self.tools.schemas()
        self._emit("\nassistant> ")
        if self.stream:
            for chunk in self.backend.stream(messages, schemas):
                self._emit(chunk)
            step = self.backend.last_step()
        else:
            step = self.backend.step(messages, schemas)
            self._emit(step.text)
        self._emit("\n")
        return step

    # -- tool execution -----------------------------------------------------

    def _run_tools(self, calls: list[ToolCall]) -> list[tuple[str, ToolResult]]:
        """Gate then run each call. Read-only calls run in parallel."""
        # Partition into parallel-safe (read-only) and serial (mutating) calls,
        # preserving original order for the returned list.
        results: dict[int, tuple[str, ToolResult]] = {}
        parallel: list[tuple[int, ToolCall]] = []
        serial: list[tuple[int, ToolCall]] = []
        for i, call in enumerate(calls):
            tool = self.tools.get(call.name)
            read_only = bool(tool and tool.read_only)
            (parallel if read_only else serial).append((i, call))

        # Run read-only calls concurrently.
        if parallel:
            with ThreadPoolExecutor(max_workers=min(4, len(parallel))) as pool:
                futures = {pool.submit(self._gate_and_run, c): i for i, c in parallel}
                for fut, i in futures.items():
                    results[i] = fut.result()

        # Run mutating calls one at a time.
        for i, call in serial:
            results[i] = self._gate_and_run(call)

        return [results[i] for i in range(len(calls))]

    def _gate_and_run(self, call: ToolCall) -> tuple[str, ToolResult]:
        tool = self.tools.get(call.name)
        read_only = bool(tool and tool.read_only)
        allowed, reason = self.gate.check(call.name, call.args, read_only)
        self._emit(f"  · {call.name}({_fmt_args(call.args)}) — {reason}\n")
        if not allowed:
            return call.name, ToolResult(f"Permission denied: {reason}", is_error=True)
        result = self.tools.call(call.name, call.args)
        # Echo a short preview of the result for the transcript.
        preview = result.content.splitlines()[0] if result.content else ""
        self._emit(f"    {preview[:100]}\n")
        return call.name, result


def _fmt_args(args: dict[str, Any]) -> str:
    parts = []
    for k, v in args.items():
        s = str(v).replace("\n", "\\n")
        parts.append(f"{k}={s[:40]}")
    return ", ".join(parts)


# ---------------------------------------------------------------------------
# REPL
# ---------------------------------------------------------------------------

def repl(agent: Agent) -> None:  # pragma: no cover - interactive
    """A minimal interactive loop: read a task, run it, repeat."""
    print("mini coding agent — type a task, or 'quit' to exit.")
    while True:
        try:
            task = input("\nyou> ").strip()
        except (EOFError, KeyboardInterrupt):
            print()
            break
        if task.lower() in {"quit", "exit"}:
            break
        if task:
            agent.run(task)


if __name__ == "__main__":  # pragma: no cover
    import tempfile

    from permissions import PermissionConfig, Mode, auto_approve

    with tempfile.TemporaryDirectory() as d:
        a = Agent(
            backend=MockBackend(),
            tools=ToolRegistry(d),
            gate=PermissionGate(PermissionConfig(mode=Mode.DEFAULT), prompter=auto_approve),
            context=ContextManager(system=DEFAULT_SYSTEM),
        )
        repl(a)
